Refactor: Implement Service Layer, Validation, and Backend Improvements

MajdaJS opened the pull request:
# Pull Request

 ## Description

This PR completes the backend refactor by introducing a clear separation between controllers and services, improving validation, and ensuring consistent error handling across all routes.

 ## Main changes:

Introduced dedicated service layer for:
authService – register, login, refresh, logout
projectsService – full CRUD with user ownership check
taskService – full CRUD linked to projects
Updated all controllers to delegate logic to their respective services
Added Zod validation schemas and integrated middleware
Unified error handling via AppError and centralized errorHandler
Enhanced logger output (structured warnings and errors)
Verified that all Jest tests (22 total) pass successfully
Added backend-node-summary.txt and updated README documentation

 ## Impact

This refactor improves code readability, testability, and maintainability.
Controllers are now lean and follow best practices for layered architecture.

qsoftdevelopment commented on backend/node/src/middleware/errorHandler.ts:
```diff
@@ -0,0 +1,35 @@
+import { Request, Response, NextFunction } from 'express';
+import { AppError } from '../utils/AppError';
+import { logger } from '../utils/logger';
+
+export const errorHandler = (
+  err: unknown,
+  _req: Request,
+  res: Response,
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  _next: NextFunction
+) => {
+  logger.error(`❌ Error caught by handler: ${JSON.stringify(err)}`);
+
```
@MajdaJS odlicno!

Ako je odgovor već poslan (npr. neki middleware prije ovog je već poslao res.json()), onda novi pokušaj slanja odgovora baca grešku:
Error: Cannot set headers after they are sent to the client

Zato stavi:

if (res.headersSent) {
  return next(err);
}

Ovo sprječava duplo slanje odgovora i prekida daljnje procesiranje.

=====
Takodjer vidi, ako logiraš JSON.stringify(err), izgubiš stack trace i često dobiješ {} jer Error objekti po defaultu nisu JSON-serializable.

qsoftdevelopment commented on backend/node/src/utils/logger.ts:
```diff
@@ -0,0 +1,33 @@
+import fs from 'fs';
+import path from 'path';
+
+const logDir = path.join(process.cwd(), 'logs');
+
+if (!fs.existsSync(logDir)) {
+  fs.mkdirSync(logDir);
+}
```
@MajdaJS jedan security savjet i razmisljanje What If

Imamo tri ozbiljna problema:

1. Race condition – ako dva procesa istovremeno pokušaju stvoriti folder, dobit ćeš EEXIST error.
2. Nema error handlinga – ako proces nema permisije, aplikacija se ruši bez korisne poruke.
3. Nema recursive: true – ako direktorij logs ne postoji u nekom parent pathu, mkdirSync baca grešku.


"što ako deployment environment nije isti kao moj lokalni". ;)

qsoftdevelopment commented:
@MajdaJS jos par savjeta.


1. updateMany ne vraca objekt koji ocekujes
backend/node/src/services/projects.service.ts (linija 42-49)
typescriptasync update(userId: number, projectId: number, name: string) {
  const updated = await prisma.project.updateMany({
    where: { id: projectId, userId },
    data: { name },
  });
  if (updated.count === 0) throw new AppError('Project not found or not yours', 404);
  
  return updated; // ❌ ovo vraca { count: 1 }, ne project objekt
}
Problem:
updateMany vraca samo { count: number }, a ne azurirani project. Ako frontend ocekuje project objekt sa id, name, createdAt, dobit ce samo { count: 1 }.


2.  Validacija prima pogresne tipove podataka
backend/node/src/validators/projectSchemas.ts (linija 9-13)

Ova validacija propusta:

{ name: 123 } – broj umjesto stringa
{ name: null } – null umjesto stringa
{ name: ['test'] } – array umjesto stringa

preprocess pretvara sve u string bez provjere tipa.

qsoftdevelopment commented:
@MajdaJS na kraju jedan tip gdje ti log file moze narasti na par GB i napuniti disk servera. veoma cest error na serveru. tada se sruse svi procesi i sve aplikacije na server prestanu raditi. 

Logger nema rotaciju logova
backend/node/src/utils/logger.ts (linija 11)
typescriptconst logStream = fs.createWriteStream(logFile, { flags: 'a' });


Za sada, ostavi ovako. Zapamti da ces ovo trebati refaktorirati kad budete radili deployment blok (tjedan 10-11).
PS. dodaj //TODO

MajdaJS commented:
@qsoftdevelopment Implemented all mentor feedback:

Replaced updateMany with update to return full object
Unified response format across all controllers
Added return to all response calls to prevent duplicate headers
Integrated stricter Zod validation (no type coercion)
Added res.headersSent check in error handler
Adjusted Jest tests for new validation and response format
Added // TODO for log rotation in logger.ts (Week 10–11)

All tests passing ✅ – backend ready for frontend integration.
